src/Simulation/QSharpFoundation/Convert/Convert.qs

```diff
@@ -63,29 +63,76 @@ namespace Microsoft.Quantum.Convert {
     /// Converts a given big integer to an equivalent integer, if possible.
     /// The function returns a pair of the resulting integer and a Boolean flag
     /// which is true, if and only if the conversion was possible.
-    /// # Remarks
-    /// See [C# BigInteger constructor](https://docs.microsoft.com/dotnet/api/system.numerics.biginteger.-ctor?view=netframework-4.7.2#System_Numerics_BigInteger__ctor_System_Int64_) for more details.
     function MaybeBigIntAsInt(a : BigInt) : (Int, Bool) {
-        body intrinsic;
+        if a > (1L <<< 63) - 1L or a < (-1L * (1L <<< 63)) {
+            return (0, false);
+        }
+        let arr = BigIntAsBoolArray(a);
+        let len = Length(arr);
+
+        // BigIntAsBoolArray always returns padded results with minimum length 8, so the below
+        // logic can assume the last entry is the sign bit for two's complement.
+        mutable val = 0;
+        for i in 0..(len - 2) {
+            set val += arr[i] ? 2 ^ i | 0;
+        }
+        if arr[len - 1] {
+            set val -= 2 ^ (len - 1);
+        }
+
+        return (val, true);
     }
 
 
     /// # Summary
     /// Converts a given big integer to an array of Booleans.
     /// The 0 element of the array is the least significant bit of the big integer.
-    /// # Remarks
-    /// See [C# BigInteger constructor](https://docs.microsoft.com/dotnet/api/system.numerics.biginteger.-ctor?view=netframework-4.7.2#System_Numerics_BigInteger__ctor_System_Int64_) for more details.
     function BigIntAsBoolArray(a : BigInt) : Bool[] {
-        body intrinsic;
+        // To use two's complement, little endian representation of the integer, we fisrt need to track if the input
+        // is a negative number. If so, flip it back to positive and start tracking a carry bit.
+        let isNegative = a < 0L;
+        mutable carry = isNegative;
+        mutable val = isNegative ? -a | a;
+
+        mutable arr = [];
+        while val != 0L {
+            let newBit = val % 2L == 1L;
+            if isNegative {
+                // For negative numbers we must invert the calculated bit, so treat "true" as "0"
+                // and "false" as "1". This means when the carry bit is set, we want to record the
+                // calculated new bit and set the carry to the opposite, otherwise record the opposite
+                // of the calculate bit.
+                if carry {
+                    set arr += [newBit];
+                    set carry = not newBit;
+                }
+                else {
+                    set arr += [not newBit];
+                }
+            }
+            else {
+                // For positive numbers just accumulate the calculated bits into the array.
+                set arr += [newBit];
+            }
+
+            set val /= 2L;
+        }
```

Comment:
1. I have an impression that `BigIntAsBoolArray()` does `BitInt` arithmetic. I'm not sure how `BigInt` arithmetic works/implemented in Q# (C# runtime and QIR runtime).
L95: `-a`,
L95: I assume that `val` is of type `BitInt`,
L98: compares `BitInt` to `0`,
L118: does `BigInt` division.


2. If we use `BitInt` arithmetic here in `BigIntAsBoolArray()`, then why don't we use it in `MaybeBigIntAsInt()`? Something like this:
```
    function MaybeBigIntAsInt(a : BigInt) : (Int, Bool) {
        if a > (1L <<< 63) - 1L or a < (-1L * (1L <<< 63)) {  // Same as before.
            return (0, false);
        }

        mutable val = 0;  // Difference starts.
        mutable len = 32;
        while (len) {
            set val = val << 1; // Shift left.
            set val = val + ((a &&& (1L << 31)) >> 31); // The bit 31 of `a` is added to bit 0 of `val`.
            a = a << 1;  // Shift left.
            --len;  // Decrement `len`.
        }
        return (val, true);
```

Reply:
The problem I'm trying to avoid is the issue of arithmetic shift right vs logical shift right. Your logic might work even accounting for that, I'd need to double check.